Initial deprecations for Labels API

jni opened the pull request:
# References and relevant issues

Initial work towards resolving #6491.

# Description

This deprecates the following attributes in the Labels layer:

- num_colors: `layer.num_colors` becomes `len(layer.colormap)`. `layer.num_colors = n` becomes `layer.colormap = label_colormap(n)`.
- `napari.utils.colormaps.LabelColormap` is deprecated and has been renamed to `napari.utils.colormaps.CyclicLabelColormap`.
- color: `layer.color` becomes `layer.colormap.color_dict`. `layer.color = color_dict` becomes `layer.colormap = DirectLabelColormap(color_dict)`.
- _background_label: `layer._background_label` is now at `layer.colormap.background_value`.
- color_mode: `layer.color_mode` is set by setting the colormap using the corresponding colormap type (`CyclicLabelColormap` or `DirectLabelColormap`).
- `seed_rng`: this was never released, it's just removed.
- `seed`: this was only used for shifting labels around in [0, 1]. It was superseded by #6460 and is deprecated.

jni committed:
Deprecate Labels.num_colors in favor of setting colormap

Czaki commented on napari/layers/labels/labels.py:
```diff
@@ -562,15 +561,34 @@ def _set_colormap(self, colormap):
+            ),
+            FutureWarning,
+            stacklevel=2,
+        )
```
```suggestion
        warnings.warn(
            trans._(
                'Labels.num_colors is deprecated since 0.4.19 and will be '
                'removed in 0.5.0, please use Labels.colormap.num_colors '
                'instead.',
                deferred=True,
            ),
            FutureWarning,
            stacklevel=2,
        )
```

Czaki replied:
I do not like this -1. It is technical detail. We should define property (make `_num_unique_colors` property public.

jni replied:
Actually, the -1 is outdated — if you look below, there's no -1. I was just confused because we use different conventions in different places. I think in the future we should harmonize on one definition or the other (I'd say probably including all colors, since then it matches the length of the colorarray) and keep len, which is intuitive. For now I will just update the message to just `len(Label.colormap)` (no -1).

Czaki replied:
```python
len(viewer.layers[0].colormap)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 len(viewer.layers[0].colormap)

TypeError: object of type 'LabelColormap' has no len()
```

jni replied:
Yes you need this PR for that 😉

Czaki replied:
The current implementation will not work with direct colormap.

jni replied:
Interesting point! I'll think about ways around that... It would be easy to add a len to the direct colormap but then it is a behaviour change. But it's also a bit weird to ask for num_colors while in direct mode. 🤔

Czaki replied:
And all this problems comes from joining both colormap under one name. 

Maybe t is better to deprecate `colormap` property in Labels layer, and introduce `cyclic_colormap` and `direct_colormap` ones with keeping `color_mode` property?

jni replied:
Question: would you expect len(a_direct_colormap) to give you the number of colors or the number of keys? Colors is of course much easier to implement and I'm inclined to just do that. But I'm willing to hear counter-arguments. 😊

Czaki replied:
Keys is much easier to implement (`len(dict)`) than colors (`len(set(tuple(x) for x in dict.values()))`) 

I expect that if I run code that calls `len(Labels.colormap)` it should not require adding `if isinstance(Labels.colormap, LabelsColormap)` to not get some stupid values.

jni replied:
> Keys is much easier to implement (`len(dict)`) than colors (`len(set(tuple(x) for x in dict.values()))`)

@Czaki no, the point is that actually, it's already implemented for the values, because I added `len` to `Colormap`, not even `LabelsColormapBase`, as `len(self.colors)`. Since you construct `DirectLabelColormap` by making the above set, it already returns the number of colors. Since the answer is accurate (this is in fact the number of colors), I'm inclined to just leave it as-is... For example, if one does a mapping of 1M labels to Viridis discretised to uint8, I think 256 is a reasonable answer.

I guess the question is why people would be reading `layer.num_colors` if they are using direct mode... My guess is that this doesn't happen. (And if it does, we will hear about it as a bug report during the deprecation period (0.4.19 -> 0.5.0).)

jni replied:
(Hmmm, so do you think I should return `len(self._random_colormap)` in the deprecated implementation? Then people can notice that the results are different when they implement the suggestion.)

Czaki replied:
> I guess the question is why people would be reading `layer.num_colors` if they are using direct mode... My guess is that this doesn't happen. (And if it does, we will hear about it as a bug report during the deprecation period (0.4.19 -> 0.5.0).)

Because some plugin creators will not think earlier and forget that under colormap it could also be direct one. 

> Since you construct `DirectLabelColormap` by making the above set, it already returns the number of colors

`colors` in `DirecLabelColorap` will have length 1.

https://github.com/napari/napari/blob/d507dd64ed6d96e984fa7fa57f2f232be388b5f7/napari/utils/colormaps/colormap.py#L353-L356

In a direct colormap the colors are stored in `color_dict` attribute. 

It i not my design. I just modified the existing one. 

> do you think I should return `len(self._random_colormap)` in the deprecated implementation

For me, it will be a number of keys. But without strong opinion.

jni replied:
@Czaki I went with passing through `_num_unique_colors` because it allows unifying a few code paths between direct and auto color modes.

Czaki commented on napari/layers/labels/labels.py:
```diff
@@ -562,15 +561,34 @@ def _set_colormap(self, colormap):
     def num_colors(self, num_colors):
+        warnings.warn(
+            trans._(
+                'Setting Labels.num_colors is deprecated since 0.4.19 and '
+                'will be removed in 0.5.0, please set Labels.colormap '
+                'instead.',
```
Maybe point to `napari.utils.colormaps.label_colormap` function?

jni replied:
I think I want to discourage/remove that also in the future — we should be able to use the LabelColormap constructor with a list of colors. Maybe we make a `low_discrepancy_colors` function to generate it and then it's `LabelColormap(low_discrepancy_colors(5))`. What do you think?

Czaki replied:
ok

Czaki replied:
And such a function should be part of this PR to allow switching to the new API.

Czaki replied:
So what here?

jni committed:
Update deprecation message for num_colors

jni committed:
Remove Labels._background_label

This can be easily accessed as Labels.colormap.background_value

jni committed:
Fix failing _background_label test

jni committed:
Deprecate color_mode

jni committed:
Override __len__ for DirectLabelColormap

jni committed:
DirectLabelColormap.__len__ is number of colors

This simplifies code for using the len to e.g. find the needed texture dtype to display those colors, which can now have a single code path.

jni committed:
Remove color mode from the GUI

jni committed:
Stop using color mode internally

Czaki committed:
draft of move intensity minix level up

jni committed:
Split VispyImageLayer into Base and Image

jni committed:
Fix broken colormap shuffle

jni committed:
Remove seed_rng

jni committed:
Initial work to add colormap controls to labels layer

Czaki committed:
Merge remote-tracking branch 'napari/main' into labels-api-deprecations

Czaki committed:
fixing tests

Czaki committed:
fix more test, restore color mode switch

Czaki committed:
fix examples

Czaki committed:
improve documentation

jni commented on napari/_qt/layer_controls/_tests/test_qt_layer_controls.py:
```diff
@@ -156,6 +167,16 @@ def _create_layer_controls(layer_type_with_data):
     ],
+    ids=[
+        "labels_with_color",
+        "labels",
```
maybe:

```suggestion
        "labels_with_direct_colormap",
        "labels_with_auto_colormap",
```
?

Czaki committed:
Update napari/_qt/layer_controls/_tests/test_qt_layer_controls.py

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

jni committed:
remove seed_rng from docstring

jni committed:
Finish sentence for colormap in Labels docstring

jni committed:
Add notes to deprecated attributes in Labels docstring

Czaki committed:
Merge remote-tracking branch 'napari/main' into labels-api-deprecations

jni commented on napari/utils/colormaps/colormap.py:
```diff
@@ -236,13 +243,27 @@ def _selection_as_minimum_dtype(self, dtype: np.dtype) -> int:
+    background_value : int
+        Which value should be treated as a background
+        and mapped to transparent color.
+    interpolation : Literal['zero']
+        required by implementation, please do not set value
```
Can we somehow freeze this value, rather than requiring users to cooperate?

aganders3 replied:
```diff
diff --git a/napari/utils/colormaps/colormap.py b/napari/utils/colormaps/colormap.py
index d48bcffa3..d43a984de 100644
--- a/napari/utils/colormaps/colormap.py
+++ b/napari/utils/colormaps/colormap.py
@@ -176,7 +176,7 @@ class LabelColormapBase(Colormap):
     background_value: int = 0
     interpolation: Literal[
         ColormapInterpolationMode.ZERO
-    ] = ColormapInterpolationMode.ZERO
+    ] = Field(ColormapInterpolationMode.ZERO, frozen=True)
     _cache_mapping: Dict[Tuple[np.dtype, np.dtype], np.ndarray] = PrivateAttr(
         default={}
     )
```
Does this do what you'd like? With this change, setting the field to anything else would raise a `ValidationError`:
```
ValidationError: 1 validation error for LabelColormap
interpolation
  unexpected value; permitted: <ColormapInterpolationMode.ZERO: 'zero'> (type=value_error.const; given=linear; permitted=(<ColormapInterpolationMode.ZERO: 'zero'>,))
  ```

jni commented on napari/utils/colormaps/colormap.py:
```diff
@@ -236,13 +243,27 @@ def _selection_as_minimum_dtype(self, dtype: np.dtype) -> int:
+        and mapped to transparent color.
+    interpolation : Literal['zero']
+        required by implementation, please do not set value
+    seed : float
+        seed used for random color generation. Used for reproducibility.
+        It will be removed in the future release.
```
This is only used in the labels layer to store the (deprecated) seed, maybe we should remove it in this PR?

jni replied:
GH code search returns 0 hits of user code using `colormap.seed`. So I would vote for removing this directly.

Czaki replied:
This seed is a parameter of `label_colormap` function also that is part of the public API for creating colormap. 

The general story behind this change is that until this PR, the colormaps were not part of public API documentation. I try to set up [`autodoc-pydantic`](https://pypi.org/project/autodoc-pydantic/), but it do not work out of the box. So I decided to document missing parameters. 

As we miss many docstrings in our codebase, I prefer to clearly write that argument is deprecated, than left it undocumented.

jni replied:
I'm just saying it's not used even after it is stored. So maybe we just remove it from the class, not leave it undocumented. Yes it is a public API but I don't think anyone is using it (and if they are they have plenty of recourse to fix in their own code). But, maybe we can do that in 0.5.0, is that what you would prefer?

jni committed:
Minor docstring update

jni committed:
Freeze interpolation mode on LabelColormapBase

brisvag commented on napari/layers/labels/labels.py:
```diff
@@ -287,22 +288,20 @@ def __init__(
         rendering='iso_categorical',
         rotate=None,
         scale=None,
-        seed_rng=None,
```
is this supposed to be gone like this with no warning?

Czaki replied:
Yes. As it was introduced after 0.4.18 release.

brisvag commented on napari/layers/labels/labels.py:
```diff
@@ -311,14 +310,14 @@ def __init__(
-        self._num_colors = num_colors
         self._random_colormap = label_colormap(
-            self.num_colors, self.seed, self._background_label
+            49, self._seed, background_value=0
```
maybe set this `49` to a variable and/or add a comment so we don't lose track of the origin and reason.

jni replied:
Just gonna add a comment. I think this is minor and may change in the future (e.g. with a default colormap name) so I don't think it warrants its own variable. 😅

brisvag reviewed (commented):
Generally all good, left a couple comments!

jni committed:
Add comment to explain 49 colors magic number

jni committed:
Improve deprecation messages

jni committed:
Merge branch 'main' into labels-api-deprecations

Czaki commented on napari/layers/labels/labels.py:
```diff
@@ -972,14 +1006,6 @@ def _calculate_contour(
-                raw_dtype,
-            )
-        return _texture_dtype(self.num_colors, raw_dtype)
-
```
this removal leads to current crash

jni committed:
Re-add helper method accidentally removed in merge

Czaki committed:
fix test

brisvag commented on napari/layers/labels/labels.py:
```diff
@@ -1001,6 +1007,9 @@ def _calculate_contour(
 
+    def _get_cache_dtype(self, raw_dtype: np.dtype) -> np.dtype:
+        return _texture_dtype(len(self.colormap), raw_dtype)
+
```
is this temporary to fix a bug and then should go away? If so, maybe put a comment so it's easy to find when we complete the deprecation.

brisvag reviewed (approved):
Still looks good in general :)

brisvag reviewed (approved):
Tok a higher level look at the API changes, and I agree with everything. One note: does `new_colormap` need to exist on the public api? It seems like it serves similar purpose as the old `seed`, maybe it should be private or exposed on the colormap itself?

jni committed:
Rename labels colormap classes

jni committed:
Disable color mode combobox when no direct map is set

jni committed:
Add function for deprecating class names

jni committed:
Add deprecated LabelColormap name

jni committed:
Add LabelColormap to the colormap utils init

jni committed:
Update class deprecation code to fix typing and names

psobolewskiPhD commented on napari/layers/labels/labels.py:
```diff
@@ -555,26 +552,45 @@ def _set_colormap(self, colormap):
+        warnings.warn(
+            trans._(
+                'Labels.num_colors is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0, please use len(Labels.colormap) '
```
I'm all for removing `num_colors`, but kind hate the alternative being `len(...)`, because of the lack of suggest/complete using IDE/ipython/console... 
What about having `num_colors` on the Colormap object? So one could do: `Labels.colormap.num_colors`

jni replied:
@psobolewskiPhD maybe we can discuss this in a follow-up issue or PR? I am of the opposite opinion, because we have way too many attributes in our objects — frequently when using napari I do `obj.<TAB>` and then I get an enormous list in which it is really hard to find what I'm looking for. I'm not sure whether colormaps fall in this category, but the point is that *eventually* all objects land in this category if you keep expanding the API for every thing that you want to know about. And, it's also real easy to keep adding things, but hard to remove them, so, especially while we are working to simplify our APIs, for every attribute we add to a class, I want to have a nice, long discussion about it. 😅 For standard Python things like "how many things are in this collection", I prefer to use `len()`. I realise that not everyone would think to do this, which is why I'm happy to contemplate in a follow-up issue, but since adding to the API is easy and not time-sensitive, I suggest we leave the discussion until post-0.5.

psobolewskiPhD replied:
fair enough

psobolewskiPhD commented on napari/layers/labels/labels.py:
```diff
@@ -97,9 +97,9 @@ class Labels(_ImageBase):
-    color : dict of int or None to str or array
-        Custom label to color mapping. Values must be valid color names or RGBA
-        arrays. None is used when no color is specified for a label.
+    colormap : CycleLabelColormap or DirectLabelColormap or None
```
I like this change from color to colormap, makes it more intuitive.
Any thought to `CyclicLabelColormap` vs `CycleLabelColormap` ? I prefer the first, but not strongly.

jni replied:
I like that now the two class names are the same length. 😊 Done in 6f340e8

psobolewskiPhD commented on napari/layers/labels/labels.py:
```diff
@@ -189,9 +186,11 @@ class Labels(_ImageBase):
         the layer will render incorrectly if they map to more than 1024 distinct
-        colors.
+        colors. DEPRECATED: set ``colormap`` directly, using
+        `napari.utils.colormaps.direct_colormap`.
```
```suggestion
        `napari.utils.colormaps.direct_colormap` or `napari.utils.colormaps.label_colormap`
```

jni replied:
No, here we are replacing setting color, which is the direct_colormap. For num_colors, users would use label_colormap.

jni replied:
(However, I've changed the pointer to the class rather than the helper function.)

psobolewskiPhD commented on napari/layers/labels/labels.py:
```diff
@@ -279,6 +278,8 @@ class Labels(_ImageBase):
 
     _history_limit = 100
 
+    @deprecated_constructor_arg_by_attr("color")
```
I tried to pass `color=dict_of_colors` to `add_labels` and got
`TypeError: add_labels() got an unexpected keyword argument 'color'`
I was expecting to get a deprecation warning...

psobolewskiPhD replied:
There are also two uses of `color=` in:
`benchmarks/benchmark_labels_layer.py`
line 101 and 135

Czaki replied:
Will fix the problem with `add_labels tomorrow

jni replied:
@Czaki happy to take this on if you point me to how I should fix this...

psobolewskiPhD commented on napari/layers/labels/labels.py:
```diff
@@ -785,17 +796,41 @@ def color_mode(self):
+            trans._(
+                'Labels.color_mode is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0. Please check type(Labels.colormap) '
+                'instead. napari.utils.colormaps.LabelColormap corresponds to '
```
```suggestion
                'instead. napari.utils.colormaps.CycleLabelColormap corresponds to '
```
Right? LabelColormap is the thing being deprecated?

psobolewskiPhD commented on napari/layers/labels/labels.py:
```diff
@@ -785,17 +796,41 @@ def color_mode(self):
+            trans._(
+                'Labels.color_mode is deprecated since 0.4.19 and will be '
+                'removed in 0.5.0. Please set Labels.colormap instead, to an'
+                'instance of napari.utils.colormaps.LabelColormap for "auto"'
```
```suggestion
                'instance of napari.utils.colormaps.CycleLabelColormap for "auto"'
```

psobolewskiPhD commented on examples/add_labels_with_features.py:
```diff
@@ -56,7 +58,7 @@
     name='segmentation',
     features=label_features,
-    color=color,
+    colormap=DirectLabelColormap(color_dict=colors),
```
This example ends up looking different than in the gallery.
With this PR I see:
![image](https://github.com/napari/napari/assets/76622105/bd714394-a23f-4f5c-9388-1448a4636328)

with main:
![image](https://github.com/napari/napari/assets/76622105/e950818f-e76b-4678-8eac-5d4604607bde)

I'm not which is more correct, because I'm not sure if I get this example. I'd expect the large coins to get one color and the small ones to get a different color.

psobolewskiPhD replied:
I guess if None is to be magenta, then the new way is technically correct?

Czaki replied:
Yes. The old behavior was adding transparent color for the background. There is a need to add `0: 'transparent'` to color dict.

jni replied:
I think we should maybe add the background = 0 behaviour as default to the DirectLabelColormap to preserve the old behaviour (and this is the most intuitive behaviour for our users, I think). Then it could be disabled by passing `background_value=None` to the constructor. Thoughts?

Czaki replied:
I do not like it. It complicates API. But we could to this with `direct_colormap` helper function https://github.com/napari/napari/blob/3c58ab41e1a1772f133852a6c6c4adf7e209a576/napari/utils/colormaps/colormap_utils.py#L567-L585

jni replied:
Ah, I see that the background value *is* 0 by default. Well, I think by default the background should always be fully transparent — that is, if there is not a key for the background value, we add it. @Czaki?

jni replied:
I've implemented this in ddbe9c8 (with a slight tweak to the docstring in 8709db7)

psobolewskiPhD commented on napari/components/_tests/test_prune_kwargs.py:
```diff
@@ -29,7 +28,6 @@
         'labels',
         {
             'scale': (0.75, 1),
-            'num_colors': 10,
```
The docstring of this function in `viewer_model.py` also uses `num_colors` in the example.

jni replied:
Fixed in df05b49

psobolewskiPhD commented on napari/utils/colormaps/colormap.py:
```diff
@@ -331,17 +353,29 @@ def shuffle(self, seed: int):
     use_selection : bool
-        Whether to color using the selected label.
+        Whether map only selected label.
+        If `True` only selected label will be mapped to not transparent color..
```
```suggestion
        Whether to map only the selected label to a color.
        If `True` only selected label will be not transparent.
```

psobolewskiPhD reviewed (changes requested):
Sorry for the delay. I like the concept overall, it's more intuitive--to me at least. 
I think I caught a few missed places for deprecations? 
Also I think having `colormap.num_colors` or such would be nicer than using `len(colormap)`

Re: naming
I'm OK with the current or using DictColormap and CyclicColormap to make them more generic.

jni committed:
Update docstring

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

jni committed:
Add background value color in color-dict validator

jni committed:
Update docstring for colordict validator

jni committed:
Update doctest for napari.components.viewer_model.prune_kwargs

jni committed:
Update deprecation warnings for color mode to use new class name

jni committed:
Rename CycleLabelColormap to CyclicLabelColoramp

jni committed:
Point directly to colormap class instead of function

jni committed:
Fix deprecated class utility function

jni committed:
ignore typing on the signature copy

Czaki committed:
proper ignore

Czaki committed:
add mechanism for deprecation of `add_'layer'`

Czaki committed:
Merge branch 'main' into labels-api-deprecations

jni commented on napari/utils/_register.py:
```diff
@@ -76,36 +93,3 @@ def create_func(cls, name=None, doc=None, filename: str = '<string>'):
     return func
-
-
-def _register(cls, *, name=None, doc=None):
-    from napari.components import ViewerModel
-
-    func = create_func(cls, name=name, doc=doc)
-    setattr(ViewerModel, func.__name__, func)
-    return cls
-
-
-def add_to_viewer(cls=None, *, name=None, doc=None):
-    """Adds a layer creation convenience method under viewers
-    as ``add_{name}``.
-
-    Parameters
-    ----------
-    cls : type, optional
-        Class to register.
-        If None, this function is treated as a decorator.
-    name : string, keyword-only
-        Name in snake-case of the layer name.
-        If None, is autogenerated from the class name.
-    doc : string, keyword-only
-        Docstring to use in the method.
-        If None, is autogenerated from the existing docstring.
-    """
-    if cls is not None:
-        return _register(cls, name=name, doc=doc)
-
-    def inner(cls):
-        return _register(cls, name=name, doc=doc)
-
-    return inner
```
Why are these deleted?

Czaki replied:
Because it is unused.

jni replied:
Seems like it should be in a separate PR but ok. 👍

Czaki committed:
fix signature

Czaki committed:
Merge remote-tracking branch 'napari/main' into labels-api-deprecations

Czaki committed:
fix tests

Czaki committed:
Merge branch 'main' into labels-api-deprecations

Czaki commented on napari/utils/colormaps/colormap.py:
```diff
@@ -350,17 +372,29 @@ def shuffle(self, seed: int):
         The dictionary mapping labels to colors.
+
     use_selection : bool
-        Whether to color using the selected label.
+        Whether to map only the selected label to a color.
+        If `True` only selected label will be not transparent.
     selection : int
         The selected label.
+    colors : ColorArray
+        Exist because of implementation details. Please do not use it.
     """
```
As background is used for validation of color_dict then it should be documented.

Czaki commented on napari/layers/labels/_tests/test_labels.py:
```diff
@@ -263,26 +285,28 @@ def test_seed():
+        with pytest.raises(
+            ValueError, match=r".*Only up to 2\*\*16=65535 colors"
+        ):
+            layer.num_colors = 2**17
```
only line triggering warning should be under `pytest.warns`. So it means that there should be multiple `with pytest.warns`

Czaki commented on napari/utils/migrations.py:
```diff
@@ -165,3 +170,44 @@ def _wrapper(*args, **kwargs):
+            warnings.warn(msg, FutureWarning, stacklevel=2)
+            if super().__new__ is object.__new__:
+                return super().__new__(cls)
+            return super().__new__(cls, *args, **kwargs)
+
+        def __init_subclass__(cls, **kwargs):
+            warnings.warn(msg, FutureWarning, stacklevel=2)
```
I think that these lines should be tested.

jni committed:
Test class name migration utility

jni committed:
Check individual lines for num_colors warnings

Czaki committed:
Merge branch 'main' into labels-api-deprecations

Czaki committed:
fix tests
